src/porepy/numerics/ad/operators.py

```diff
@@ -742,207 +816,31 @@ def _evaluate(
-                        # iteration, we need to use the original variable to get hold of
-                        # the correct dof indices, since this is the variable that was
-                        # created by the EquationSystem. However, we will tie the
-                        # indices to the id of this variable, since this is the one that
-                        # will be used for lookup later on.
-                        sub_var_known_to_eq_system: Variable = sub_var.original_variable
-                    else:
-                        sub_var_known_to_eq_system = sub_var
-
-                    # Get the index of this sub variable in the global numbering of the
-                    # EquationSystem. If an error message is raised that the variable is
-                    # not present in the EquationSystem, it is likely that this operator
-                    # contains a variable that is not known to the EquationSystem (it
-                    # has not passed through EquationSystem.create_variable()).
-                    ind_var.append(system_manager.dofs_of([sub_var_known_to_eq_system]))
-                    if i == 0:
-                        # Store id of variable, but only for the first one; we will
-                        # concatenate the arrays in ind_var into one array
-                        variable_ids.append(variable.id)
-
-                if len(variable.sub_vars) == 0:
-                    # For empty lists of subvariables, we still need to assign an id
-                    # to the variable.
-                    variable_ids.append(variable.id)
-            else:
-                # This is a variable that lives on a single grid
-                if variable.prev_iter or variable.prev_time:
-                    # If this is a variable representing a previous time step or
-                    # iteration, we need to use the original variable to get hold of
-                    # the correct dof indices, since this is the variable that was
-                    # created by the EquationSystem. However, we will tie the
-                    # indices to the id of this variable, since this is the one that
-                    # will be used for lookup later on.
-                    variable_known_to_eq_system = variable.original_variable
-                else:
-                    variable_known_to_eq_system = variable
-
-                ind_var.append(system_manager.dofs_of([variable_known_to_eq_system]))
-                variable_ids.append(variable.id)
-
-            # Gather all indices for this variable
-            if len(ind_var) > 0:
-                inds.append(np.hstack([i for i in ind_var]))
-            else:
-                inds.append(np.array([], dtype=int))
-
-        return inds, variable_ids, prev_time, prev_iter
-
-    def _find_subtree_variables(self) -> Sequence[Variable]:
+    def find_variables_in_tree(self) -> Sequence[Variable | MixedDimensionalVariable]:
```

Comment:
My motivation is that any operator should have a way to find it's independent variables, and the *set* of grids on which it is defined (the latter I did not implement for this PR, but the recursion is analogous).
I think this will be required for any kind of optimization in the operator tree.
Also, since porepy represents equations as operators, this is an easy way to find the independent variables (and domains of definition) which are part of the equation.

Reply:
Fair enough.